service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/TestPopulateCaseController.java

```diff
@@ -0,0 +1,65 @@
+package uk.gov.hmcts.reform.fpl.controllers;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.swagger.annotations.Api;
+import lombok.RequiredArgsConstructor;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+import uk.gov.hmcts.reform.ccd.client.model.AboutToStartOrSubmitCallbackResponse;
+import uk.gov.hmcts.reform.ccd.client.model.CallbackRequest;
+import uk.gov.hmcts.reform.ccd.client.model.CaseDetails;
+import uk.gov.hmcts.reform.fpl.utils.ResourceReader;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+@Api
+@RestController
+@RequestMapping("/callback/populate-case")
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+public class TestPopulateCaseController {
```

Comment:
personally i would create regular endpoint for it (not a callback) on which we can apply strict access control (see RoboticsController and SecurityConfiguration#RoboticsSecurityConfiguration, alternatively, we could also have env variable or feature toggle which blocks endpoint) We can then do a direct call from e2e to fpl service - should be exposed by ingress) and then from this endpoint we can issue ccd event (or series of events). I guess is is more flexible from test perspective as you can easely create dynamic payload

Reply:
That's another possible solution, but the question is if we need more flexibility on e2e tests. I'd say that selecting json file that should be used to populate case is enough for our current needs, so maybe it's not worth going for more complex solution. The point on isolating non prod code is worth considering, but the event I added is only available to system-update user, who can already do everything with the case data.

I was also thinking about calling ccd api from the e2e test itself. That wouldn't require any changes in the service, but I'm not sure if accessing ccd api from there is possible.

And I agree on the point of finding a way to change case state. Maybe we could trigger ccd events responsible for that in pre-populate submitted event? Or maybe we just use the populate-case event to fill case data between state changes, and then we leave it as it is now (manually clicking through the event that's changing the state)?

Reply:
I believe @toireasam is currently looking into updating [STATE] variable in case data in callback which might allow us to do easy state changing via callback

Reply:
yes looks like we can change state in callback by making a couple of tweaks like so :

caseDetails.getData().put("state", "Review");

This worked for me in about to submit but not mid event